Models compression via NNCF

LeonidBeynenson opened the pull request:
The NNCF support is transparent: if
* NNCF is not installed
* and the mmdetection config file does not have nncf-related parameters,
then mmdetection will work in the same way as earlier.

Also this pull request contains changes to make the script `tools/test_exported.py` be compatible with OpenVINO 2021.1

LeonidBeynenson committed:
Add NNCF support to mmdetection

The NNCF support is transparent: if
* NNCF is not installed
* and the mmdetection config file does not have nncf-related parameters,
then mmdetection will work in the same way as earlier.

LeonidBeynenson committed:
Remove the method BaseDetector.export

The method ran export of the model to onnx, but now the function
torch.onnx.export is run in the script tools/export.py directly
-- this is required to make export of NNCF-wrapped models.

LeonidBeynenson committed:
Add config files to make NNCF compression

The files should be refactored a bit.

LeonidBeynenson commented:
2 @Ilya-Krylov, @druzhkov-paul: please, review this pull request

LeonidBeynenson commented:
2 @aleksei-kashapov : FYI

druzhkov-paul commented on mmdet/apis/fake_input.py:
```diff
@@ -0,0 +1,33 @@
+from mmdet.parallel.data_cpu import scatter_cpu
+
+from mmdet.datasets.pipelines import Compose
+from .inference import LoadImage
```
```suggestion
from mmcv.parallel import collate, scatter

from mmdet.datasets.pipelines import Compose
from mmdet.parallel.data_cpu import scatter_cpu
from .inference import LoadImage
```

druzhkov-paul commented on mmdet/apis/train.py:
```diff
@@ -4,13 +4,16 @@
+from mmdet.core import DistEvalHook, EvalHook, Fp16OptimizerHook, CompressionHook
 from mmdet.datasets import build_dataloader, build_dataset
 from mmdet.utils import get_root_logger
 from mmdet.parallel import MMDataCPU
 
+from mmdet.core.nncf import wrap_nncf_model
+from .fake_input import get_fake_input
```
Just wondering, what's the logic (if any) behind these empty lines placement in the imports section?

LeonidBeynenson replied:
Historical reasons.    
If you suppose that these empty lines are evil, we could remove them.

druzhkov-paul commented on mmdet/apis/train.py:
```diff
@@ -78,6 +81,19 @@ def train_detector(model,
+        model = model.cuda()
+
+    # nncf model wrapper
+    if cfg.ENABLE_COMPRESSION:
```
WHY IS IT UPPER CASE?

LeonidBeynenson replied:
Historical reasons. This code was in the original NNCF patch for integration to the mmdetection -- 
[see the NNCF patch here](https://github.com/openvinotoolkit/nncf_pytorch/tree/develop/third_party_integration/mmdetection)

Most probably, just to distinguish the fields that were not in the config file indeed (but were added to the config struct later)

kshpv replied:
let's make lowercase

druzhkov-paul commented on mmdet/apis/train.py:
```diff
@@ -97,6 +113,9 @@ def train_detector(model,
         model = MMDataCPU(model)
         map_location = 'cpu'
 
+    if cfg.ENABLE_COMPRESSION and distributed:
+        compression_ctrl.distributed()
```
Is it necessary to do that *after* wrapping the model with `MMDistributedDataParallel`?

LeonidBeynenson replied:
I think yes. As far as I remember, NNCF wrapper classes have their own logic, and the case with distributed training should be additionally  handled.

kshpv replied:
Yes, it has to be _after_ wrapping the model with MMDistributedDataParallel as this function broadcasts parameters to all created processes

druzhkov-paul commented on mmdet/apis/train.py:
```diff
@@ -78,6 +81,19 @@ def train_detector(model,
             seed=cfg.seed) for ds in dataset
     ]
 
+    if cfg.load_from:
+        load_checkpoint(model=model, filename=cfg.load_from)
```
Why don't you need to handle `resume_from` option here as well?

LeonidBeynenson replied:
Ok, we will consider it (TBD)

LeonidBeynenson replied:
Done

Ilya-Krylov replied:
Where is it done? And looks like you load checkpoint twice. One time is here, and one more time is in wrap_nncf_model. Also checkpoint is loaded two times in test.py. Is it possible to wrap_nncf_model without loading weights and load weights after in current mmdetection manner?

druzhkov-paul commented on mmdet/apis/train.py:
```diff
@@ -137,11 +156,15 @@ def train_detector(model,
             dist=distributed,
             shuffle=False)
         eval_cfg = cfg.get('evaluation', {})
+        if "should_evaluate_before_run" not in eval_cfg:
```
1. Single quotes please. Everywhere.
1. Is this config option introduced by you? If so, I don't like the name.

LeonidBeynenson replied:
> 1. Single quotes please. Everywhere.

Are you sure? Is this style requirement from the original mmdetection?    
(If no, I do not like it, I prefer double quotes everywhere!)

> Is this config option introduced by you? If so, I don't like the name.

It was introduced by me. I did not invented a better version of the name -- I thought it is short and descriptive.   
   
Could you suggest another or point it's drawback?

LeonidBeynenson replied:
Fixed

druzhkov-paul commented on mmdet/core/post_processing/bbox_nms.py:
```diff
@@ -10,6 +10,7 @@
 from ..utils.misc import topk, dummy_pad
 from ...ops.nms import batched_nms
 from ...utils.deployment.symbolic import py_symbolic
+from ..nncf.utils import no_nncf_trace
```
Keep imports sorted please. Everywhere.

LeonidBeynenson replied:
Could you point to a specification on the sorting -- particularly, should `from .bbb` be before `from ...ccc` and after `from ...aaa`, or dots are counted in sorting?

Or some standard tool should be used? What about [such tool as isort](https://pypi.org/project/isort/)?

druzhkov-paul replied:
Just for the record of our offline discussion. Coding style is described in the [contribution guide](https://github.com/openvinotoolkit/mmdetection/blob/ote/.github/CONTRIBUTING.md).

druzhkov-paul commented:
@LeonidBeynenson you come to me and you say -- 'Please, review this pull request.' -- But you don't ask with respect. You don't even offer a high-level description of the changes. Instead you open the PR in the repo and you ask to complete the review ASAP.

druzhkov-paul commented on mmdet/models/detectors/base.py:
```diff
@@ -155,21 +173,41 @@ def forward(self, img, img_metas, return_loss=True, **kwargs):
+
+        self.img_metas = img_metas
+        self.forward_backup = self.forward
+        self.forward = partial(self.forward, return_loss=False, forward_export=True, img_metas=None)
```
Why don't you simply call `forward_export`, but instead hide it inside `forward` -> `forward_test` and pass extra arguments?

LeonidBeynenson replied:
As I wrote in my [big comment below](https://github.com/openvinotoolkit/mmdetection/pull/73#issuecomment-707357700), it is required to avoid replacing of `forward` method. There were some cases when replacing of the `forward` method worked incorrectly with NNCF, since NNCF uses its own tricks with `forward` method.

If we replace `forward` by `partial` of the same `forward` with "hardwired" parameters, it is a safe decision -- we do not care which `forward` method is substituted, and even if `forward` method was wrapped and substituted by somebody else. If during wrapping the keyword arguments are passed as `**kwargs` to the original `forward`, this trick will work in any case.

Ilya-Krylov commented:
@LeonidBeynenson please create corresponding PR to training_extensions so that we could see that tests are passing there as well.

druzhkov-paul commented on mmdet/models/detectors/base.py:
```diff
@@ -134,14 +147,19 @@ def forward_test(self, imgs, img_metas, **kwargs):
+        pass
+
+    @auto_fp16(apply_to=('img',))
+    def forward(self, img, img_metas=[], return_loss=True, **kwargs):
```
What's the reason to add (BTW mutable) default value for `img_metas`?

LeonidBeynenson replied:
OMG, it is a bug. Sorry, will fix it.

LeonidBeynenson replied:
Fixed

druzhkov-paul commented on mmdet/ops/nms/nms_wrapper.py:
```diff
@@ -164,7 +166,12 @@ def batched_nms(bboxes, scores, inds, nms_cfg, class_agnostic=False):
+            # NB: this trick is required to make class-separate NMS using ONNX NMS operation;
+            #     the ONNX NMS operation supports another way of class separation (class-separate scores), but this is not used here.
+            # TODO: check if it is possible in this architecture use class-separate scores that are supported in ONNX NMS.
+            bboxes_for_nms = bboxes + offsets[:, None]
```
1. You can enable another NMS export path [here](https://github.com/openvinotoolkit/mmdetection/blob/62803d0fdc91e575122305fed98bce84464b54d4/mmdet/utils/deployment/symbolic.py#L267).
1. I'd expect the comment to explain, what's the problem here from NNCF perspective, 'cause the trick is 100% valid from both PyTorch and ONNX side.

LeonidBeynenson replied:
1. Ok, we should investigate it further

2. Added more descriptive comment there

kshpv replied:
2. The problem is that without no_nncf_trace() context NNCF tries to optimize this operation. So, e.g. it inserts FakeQunatizer here. It leads to accuracy drop as this operation works with pixels coordinates which range from 0 to image_size which can be bigger than usual quantization range 0-255. So it leads to overlapping of some classes

druzhkov-paul commented on tools/export.py:
```diff
@@ -155,11 +153,13 @@ def export_to_openvino(cfg, onnx_model_path, output_dir_path, input_shape=None,
               'openvino/bin/setupvars.sh before running this script.')
         return
 
+    # command_line += ' --static_shape'
```
?

LeonidBeynenson replied:
Ok, will be removed

LeonidBeynenson replied:
Fixed

druzhkov-paul commented on tools/export.py:
```diff
@@ -224,7 +219,8 @@ def main(args):
         export_to_onnx(model, fake_data, export_name=onnx_model_path, opset=args.opset,
-                       alt_ssd_export=getattr(args, 'alt_ssd_export', False))
+                       alt_ssd_export=getattr(args, 'alt_ssd_export', False),
+                       verbose=True)
```
Why do you need it to be verbose?

LeonidBeynenson replied:
It is required to compare ONNX verbose outputs for different models (or after some changes in the code).

It does not negatively affect the user experience, just a bit more logs.)

druzhkov-paul commented on tools/test.py:
```diff
@@ -114,16 +130,27 @@ def main():
+    if cfg.get('nncf_config'):
+        check_nncf_is_enabled()
+        cfg.nncf_load_from = args.checkpoint
+        model.cuda()  # for wrap_nncf_model
```
What if CUDA is not available?

LeonidBeynenson replied:
Fixed

druzhkov-paul commented on tools/test.py:
```diff
@@ -114,16 +130,27 @@ def main():
+        model.cuda()  # for wrap_nncf_model
+        _, model = wrap_nncf_model(model, cfg, None, get_fake_input)
+        checkpoint = load_checkpoint_state_dict(args.checkpoint)
+        # FIXME: TODO: check why checkpoint does not have "meta" inside
```
Please complete this.

LeonidBeynenson replied:
Fixed

druzhkov-paul commented on tools/test.py:
```diff
@@ -14,6 +15,21 @@
+from mmdet.apis import get_fake_input
+
+def load_checkpoint_state_dict(filename):
+    # this code is the common part of mmcv.runner.load_checkpoint and nncf.load_state
```
What is wrong with `mmcv.runner.load_checkpoint`?

LeonidBeynenson replied:
Ok, I will try to use it.

LeonidBeynenson replied:
Done

druzhkov-paul commented on mmdet/core/nncf/compression.py:
```diff
@@ -0,0 +1,122 @@
+    class DummyInitializingDataLoader:
+        pass
+
+
```
```suggestion
```

druzhkov-paul commented on mmdet/core/nncf/compression.py:
```diff
@@ -0,0 +1,122 @@
+
+        class_InitializingDataLoader = InitializingDataLoader
+    except:
+        raise RuntimeError("Incompatible version of NNCF")
```
Is it possible to suggest a proper version here?

LeonidBeynenson replied:
I'm not sure that NNCF library has a stable versioning at the moment.

I will consider it with NNCF guys.

druzhkov-paul replied:
NNCF has versioned packages on PyPI. I guess they do have pretty stable versioning.

LeonidBeynenson replied:
Wrote about that in `docs/nncf_config.md`

druzhkov-paul commented on mmdet/core/nncf/compression.py:
```diff
@@ -0,0 +1,122 @@
+            model(img)
+
+    if "nncf_should_use_dummy_forward_with_export_part" in cfg:
+        # TODO: this parameter is for debugging, remove it later
```
Have you finished debugging?

LeonidBeynenson replied:
No. At the moment we are not sure if we should use "forward_dummy" or "forward_export" approach.

I hope to use "forward_export" -- in this case NNCF wrappers will be able to trace the whole model's graph.

The approach with "forward_export" is a fallback decision at the moment.
But it may be turned out (in the future) that it is even better than "forward_export" approach. But we are not sure.

To check that we should try the approach on ATSS, Faster-RCNN, Mask-RCNN, and several other type of models.

The work is in the progress.

LeonidBeynenson replied:
Changed the name of the parameter, add detailed description in comments to the code and in `docs/nncf_config.md`

druzhkov-paul commented on mmdet/core/nncf/compression.py:
```diff
@@ -0,0 +1,122 @@
+    return compression_ctrl, model
+
+
+def unwrap_nncf_model(module):
```
This is a bit confusing that `wrap_nncf_model` changes the model object, while `unwrap_nncf_model` does not. Not sure that this is something bad, though.

LeonidBeynenson replied:
Agreed. We will consider it with @aleksei-kashapov and will try to find a better name.

kshpv replied:
Good comment. I suggest to change the name from **unwrap_nncf_model** to **get_uncompressed_model**

LeonidBeynenson replied:
Done

LeonidBeynenson commented:
> @LeonidBeynenson you come to me and you say -- 'Please, review this pull request.' -- But you don't ask with respect. 

I beg your pardon, dear sir!

This PR allows to wrap mmdetection models with NNCF classes and make compression of the models.    
The official documentation of NNCF you can see in the repo    
https://github.com/openvinotoolkit/nncf_pytorch

The most of code that connects NNCF with mmdetection is placed in the folder [mmdet/core/nncf](https://github.com/LeonidBeynenson/mmdetection/tree/lb/merge_NNCF_support_to_ote/mmdet/core/nncf).     
(Please, note that this folder is placed in the folder `mmdet/core` -- it is required to import it from other parts of mmdetection)    
The main function that wraps mmdetection models into NNCF classes is `wrap_nncf_model` in the file [mmdet/core/nncf/compression.py](https://github.com/LeonidBeynenson/mmdetection/blob/lb/merge_NNCF_support_to_ote/mmdet/core/nncf/compression.py)

The main changes that we made in your repo for this integration:
1. add special parameters to the method `forward_test` of the BaseDetector.  
    So, now 
    * if we call the method `forward` of the `BaseDetector` with parameters     
      `return_loss=False`, `dummy_forward=True`, then the method `forward_dummy` is called instead of `forward_test` body
    * if we call the method `forward` of the `BaseDetector` with parameters     
      `return_loss=False`, `forward_export=True`, then the forward_export is called instead of `forward_test` body

    This approach allows to avoid replacing `forward` methods in BaseDetector itself-- it may be impossible in some cases if we wrap the model by NNCFNetwork class (whereas all the parameters are passed through NNCFNetwork class transparently)

1. Add context helpers `forward_export_context` and `forward_dummy_context` to the BaseDetector class.    
    It allows to make this addition of `return_loss=False, forward_export=True` parameters to the forward method of the model     
    We use special methods for these operations, since we need them both for ONNX tracing and for NNCF tracing (note that NNCF use its own tracer to build graph of pytorch model to insert `FakeQuantize` nodes on activations -- see the method )

1. The function `get_fake_input` is moved to a separate file `mmdet/apis/fake_input.py`, since it is used in several different script.    
    Particularly, it is used to run NNCF tracing -- see the function `wrap_nncf_model`

1. If `nncf_config` section is present in the config file, then the work with NNCF is turned on. Otherwise it is turned off (and everything should work as earlier)

LeonidBeynenson commented:
> @LeonidBeynenson please create corresponding PR to training_extensions so that we could see that tests are passing there as well.

2 @Ilya-Krylov: Done    
See https://github.com/openvinotoolkit/training_extensions/pull/346

LeonidBeynenson committed:
Reorder imports in mmdet/apis/fake_input.py

Co-authored-by: Pavel Druzhkov <pavel.druzhkov@intel.com>

LeonidBeynenson committed:
Remove some empty lines

Co-authored-by: Pavel Druzhkov <pavel.druzhkov@intel.com>

LeonidBeynenson committed:
Add comment on no_nncf_trace in NMS

druzhkov-paul commented:
run tests

Ilya-Krylov commented on mmdet/core/nncf/compression_hooks.py:
```diff
@@ -0,0 +1,26 @@
+from texttable import Texttable
```
add to requirements

druzhkov-paul replied:
I'd prefer to make this import happen only if it's actually required, i.e. if nncf is installed and used.

Also it's good to add nncf itself to the list of optional dependencies.

LeonidBeynenson replied:
Done

Ilya-Krylov commented on tools/export.py:
```diff
@@ -213,6 +200,14 @@ def main(args):
     cfg = model.cfg
     fake_data = get_fake_input(cfg, device=device)
 
+    # BEGIN nncf part
```
are these comments needed?

LeonidBeynenson replied:
I think it is ok for code decoration

druzhkov-paul commented:
@LeonidBeynenson @aleksei-kashapov documentation page briefly describing this functionality is needed. Feel free to do that as a follow up PR.

Commit:
Add cuda availability checking for test.py

LeonidBeynenson committed:
Make more smart loading of NNCF-trained models

Remove nncf_load_from field from config files, now the loading of
NNCF-compressed model is done using the standard config fields load_from
and resume_from.
Now if config file contains nncf_config section, the loading of a
checkpoint is made as follows:
* load a checkpoint (using `load_from` and `resume_from` fields)
* get metadata of the loaded checkpoint,
  * if it contains nncf metadata, then load the whole NNCF-compressed model
    from the checkpoint
  * otherwise, load as a usual (non-compressed) model and then wrap it
    by NNCF (initializating NNCF fields)

Commit:
Fix bug with a lack of compression for input tensor

LeonidBeynenson committed:
Fix the default value for img_metas parameter in BaseDetector.forward

Also some comments are added.

LeonidBeynenson committed:
Fix comments and log messages

LeonidBeynenson committed:
Fix function name unwrap_nncf_model -> get_uncompressed_model

LeonidBeynenson committed:
Make more accurate work with `import texttable`

Now the code works even if texttable is not installed.

LeonidBeynenson committed:
Move integration with nncf from mmdet/core to a separate folder

It is required since it caused cyclic depenencies in some cases
inside mmdet.core

LeonidBeynenson committed:
Fix several style issues found by flake8

LeonidBeynenson committed:
Fix several style issues found by flake8 and yapf

LeonidBeynenson committed:
Fix import's order using isort

LeonidBeynenson committed:
Replace double quotes by single quotes where it is possible

LeonidBeynenson committed:
Update README.md and write docs/nncf_compression.md

LeonidBeynenson commented:
2 @aleksei-kashapov :    
Aleksei, could you review my latest changes?

If you agree, I will ask others to re-review this PR.

druzhkov-paul commented:
@LeonidBeynenson please rebase on/merge the target branch.

LeonidBeynenson committed:
Small fix in docs

kshpv reviewed (commented):
Only one comment about above changes:
I am not sure that mmdet/integration/nncf is a good path to store NNCF-related files. I prefer the previous variant.
Also, I will take a part on work on a readme file in the next commits

LeonidBeynenson committed:
Change docs/nncf_compression.md -- update documentation on NNCF integration

LeonidBeynenson committed:
Small fix in docs/nncf_compression.md

LeonidBeynenson commented:
> @LeonidBeynenson please rebase on/merge the target branch.

Done

LeonidBeynenson commented on tools/test.py:
```diff
@@ -114,16 +115,30 @@ def main():
+        if fp16_cfg is not None:
+            wrap_fp16_model(model)
+        checkpoint = load_checkpoint(model, args.checkpoint, map_location='cpu')
+        if args.fuse_conv_bn:  # TODO: FIXME: should it be inside this 'else' branch???
+            from tools.fuse_conv_bn import fuse_module
+            model = fuse_module(model)
```
2 @Ilya-Krylov:  
Ilya, could you give us your recommendation, if this `fuse_conv` operation should be inside this `else` branch (e.g. when  NNCF is not enabled)?    
I'm not sure, since I did not investigated thoroughly how it works indeed.

Ilya-Krylov replied:
To my knowledge we never used it. But from integration perspective I think NNCF should cover this case as well.
@druzhkov-paul what do you think?

LeonidBeynenson committed:
Change loading NNCF checkpoints a bit

kshpv reviewed (commented):
May we change directory of the NNCF related codes from **integration** to **external**?

LeonidBeynenson commented:
2 @Ilya-Krylov @druzhkov-paul :   
Could you re-review this pull request?   
I made all the changes which I wanted to made in the code.
The only thing that I could do is refactoring NNCF config files (the folder `configs/nncf/`)

Ilya-Krylov commented on README.md:
```diff
@@ -19,6 +19,17 @@ Please refer to [install.md](docs/install.md) for installation and dataset prepa
 
+## Models compression
+
+This project allows to make compression of the models by [NNCF](https://github.com/openvinotoolkit/nncf_pytorch) framework.
```
allows you to make or allows making

LeonidBeynenson replied:
Sorry, Ilya, but I found sufficiently many examples with "allows to make" without "you":  
https://www.google.com/search?q=%22allows+to+make%22  
https://context.reverso.net/translation/english-russian/%22allows+to+make%22  

I'd better keep the current form of the sentence...

Ilya-Krylov replied:
https://www.grammaring.com/to-infinitive-or-gerund-advise-recommend-allow-permit-forbid-require

LeonidBeynenson replied:
Yes, I've found it already too.

Fixed in one of my last commits.

Ilya-Krylov commented on configs/nncf_compression/pascal_voc/retinanet_r50_fpn_1x_voc0712_int8.py:
```diff
@@ -0,0 +1,49 @@
+find_unused_parameters = True
+nncf_load_from = None
+
+nncf_config = {
```
probably a nit,  but dictionaries in all configs are defined via dict() rather than {}

LeonidBeynenson replied:
Ok! IMHO it is a matter of style, but it is better to have uniform style, so I'll change it

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on configs/nncf_compression/vehicle-detection/vehicle-detection-0200.py:
```diff
@@ -0,0 +1,159 @@
+# model settings
```
should not it be in training_extensions?

LeonidBeynenson replied:
Yes, sir!

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on configs/nncf_compression/vehicle-detection/vehicle-detection-0200_int8.py:
```diff
@@ -0,0 +1,185 @@
+# model settings
```
should not it be in training_extensions?

LeonidBeynenson replied:
Yes, it should!

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on configs/nncf_compression/wider_face/face-detection-0200-config.py:
```diff
@@ -0,0 +1,179 @@
+# model settings
```
should not it be in training_extensions?

LeonidBeynenson replied:
Yes, it should!

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on configs/nncf_compression/wider_face/face-detection-0200-config_int8.py:
```diff
@@ -0,0 +1,176 @@
+# model settings
```
should not it be in training_extensions?

LeonidBeynenson replied:
Yes, it should!

LeonidBeynenson replied:
Fixed

Ilya-Krylov reviewed (commented):
`tools/__init__.py` ?

LeonidBeynenson committed:
Fix style of NNCF configs

Replace `val = {key: value...}` with `val = dict(key=value,...)`

LeonidBeynenson committed:
Refactor config files from configs/nncf_compression/

LeonidBeynenson committed:
Remove config files that should be moved to OTE

LeonidBeynenson commented:
2 @Ilya-Krylov @druzhkov-paul :    
I refactored and updated config files. IMHO now this PR is ready to merge.

Commit:
Update code_style of wrap_nncf_model (#2)

LeonidBeynenson committed:
Fix comments and documentation on nncf_compress_postprocessing

Also tools/__init__.py is removed.

LeonidBeynenson commented:
2 @Ilya-Krylov, @druzhkov-paul :  
Hi Ilya and Pavel,

I'm glad to inform you that I finished work on this PR.    
(Just couple of style changes were added yesterday)    
IMHO, now this PR is ready to merge.
Please, re-review it.   

Note that the documentation on these changes are in the file `docs/nncf_compression.md`

Ilya-Krylov commented on configs/nncf_compression/pascal_voc/retinanet_r50_fpn_1x_voc0712.py:
```diff
@@ -0,0 +1,27 @@
+optimizer = dict(type='SGD', lr=0.001, momentum=0.9, weight_decay=0.0001)
+optimizer_config = dict(grad_clip=None)
+# learning policy
+# actual epoch = 3 * 3 = 9
```
why `actual epoch` here is different from comment on 13th line?

LeonidBeynenson replied:
Fixed. (Removed the comments)

(Indeed, we need a smaller number of epochs for NNCF fine-tuning)

Ilya-Krylov commented on configs/nncf_compression/pascal_voc/ssd300_voc0712_int8.py:
```diff
@@ -0,0 +1,28 @@
+total_epochs = 2
+
+work_dir = './output'
+load_from = '../original_mmdetection/mmdetection/ssd300_voc_vgg16_caffe_240e_20190501-7160d09a.pth'
```
?

LeonidBeynenson replied:
2 @Ilya-Krylov:   
Ilya, could you give me a proper link where could I download the model from mmlab?
I found only a mirror:
https://s3.ap-northeast-2.amazonaws.com/open-mmlab/mmdetection/models/ssd300_voc_vgg16_caffe_240e_20190501-7160d09a.pth

Or this is the proper link itself?

LeonidBeynenson replied:
As far as I found out, this is a proper link itself. 

So, fixed in one of my last commits.

Ilya-Krylov commented on configs/nncf_compression/pascal_voc/retinanet_r50_fpn_1x_voc0712_magnitude_sparsity_int8.py:
```diff
@@ -0,0 +1,40 @@
+
+work_dir = './output'
+
+find_unused_parameters = True
```
why is it True?

LeonidBeynenson replied:
It is required by some NNCF tricks related to `DistributedDataParallel`. It was requirement from the initial NNCF patch for the original mmdetection.

Ilya-Krylov commented on docs/nncf_compression.md:
```diff
@@ -0,0 +1,136 @@
+
+## General description
+
+OTEDetection allows to make compression of the models by
```
allows making

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on docs/nncf_compression.md:
```diff
@@ -0,0 +1,136 @@
+OTEDetection allows to make compression of the models by
+[NNCF (Neural Network Compression Framework)](https://github.com/openvinotoolkit/nncf_pytorch).
+
+The work of OTEDetection is checked with [**NNCF 1.4.1**](https://github.com/openvinotoolkit/nncf_pytorch/tree/v1.4.1).
```
Should not it be added this to requirements/compression.txt ?

LeonidBeynenson replied:
Added `requirements/nncf_compression.txt`. 

Also added text about it into `docs/install.md` and `docs/nncf_compression.md`

Ilya-Krylov commented on docs/nncf_compression.md:
```diff
@@ -0,0 +1,136 @@
+The work of OTEDetection is checked with [**NNCF 1.4.1**](https://github.com/openvinotoolkit/nncf_pytorch/tree/v1.4.1).
+Please, use this version of NNCF in case of any issues.
+
+NNCF allows to make compression using the following methods
```
allows making

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on docs/nncf_compression.md:
```diff
@@ -0,0 +1,136 @@
+The work of OTEDetection is checked with [**NNCF 1.4.1**](https://github.com/openvinotoolkit/nncf_pytorch/tree/v1.4.1).
+Please, use this version of NNCF in case of any issues.
+
+NNCF allows to make compression using the following methods
```
different from what we can read in README.md. It might confuse a potential user.

LeonidBeynenson replied:
Fixed. Updated text in the main `README.md` and in `docs/nncf_compression.md`.

Ilya-Krylov commented on docs/nncf_compression.md:
````diff
@@ -0,0 +1,136 @@
+```python
+nncf_config = {
+    "input_info": {
+        "sample_size": [1, 3, 1000, 600]
````
please change to dict() view

Ilya-Krylov commented on docs/nncf_compression.md:
```diff
@@ -0,0 +1,136 @@
+2. The model is fine-tuned with compression (the parameter `nncf_config` is set).  
+   The result of this step is a checkpoint with the compressed model
+
+3. (optional) The compression precess may be resumed: model may be fine-tuned with compression a bit more
```
typo

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on mmdet/apis/train.py:
```diff
@@ -137,11 +155,15 @@ def train_detector(model,
             shuffle=False)
         eval_cfg = cfg.get('evaluation', {})
+        if 'should_evaluate_before_run' not in eval_cfg:
+            eval_cfg['should_evaluate_before_run'] = nncf_enable_compression
```
I would propose not to use this strange name/variable at all. Please create child classes for DistEvalHook and EvalHook and implement `before_run` methods and register this hooks if it necessary one line below.

As you maybe know we are going to periodically merge upstream, and we would like to minimize merging issues.

LeonidBeynenson replied:
Ok. I'll fix it in the next commit.

LeonidBeynenson replied:
Fixed.

Implemented `DistEvalBeforeRunHook` and `EvalBeforeRunHook` instead.

Ilya-Krylov commented on mmdet/integration/nncf/compression.py:
```diff
@@ -0,0 +1,165 @@
+        from nncf.nncf_network import NNCFNetwork
+
+        class_InitializingDataLoader = InitializingDataLoader
+    except:  # noqa: E722
```
is it hard to determine exact exception class?

LeonidBeynenson replied:
It is `ImportError`.

Fixed in one of my last commits.

Ilya-Krylov commented on mmdet/integration/nncf/compression.py:
```diff
@@ -0,0 +1,165 @@
+            '-- most probably, incompatible version of NNCF. '
+            'Please, use NNCF version pointed in the documentation.')
+else:
+
```
extra line

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on mmdet/integration/nncf/utils.py:
```diff
@@ -0,0 +1,92 @@
+        from nncf.dynamic_graph.context import get_current_context
+        from nncf.dynamic_graph.context import \
+            no_nncf_trace as original_no_nncf_trace
+    except:  # noqa: E722
```
ditto

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on mmdet/integration/nncf/utils.py:
```diff
@@ -0,0 +1,92 @@
+            'Please, use NNCF version pointed in the documentation.')
+
+
+def load_checkpoint(model, filename, map_location=None, strict=False):
```
what's wrong with load_checkpoint from mmcv?

Ilya-Krylov commented on mmdet/models/detectors/base.py:
```diff
@@ -112,7 +114,20 @@ def forward_test(self, imgs, img_metas, **kwargs):
                 augs (multiscale, flip, etc.) and the inner list indicates
                 images in a batch.
+
+        Note that if `kwargs` contains either `forward_dummy=True` or
```
forward_export ?

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on tools/test.py:
```diff
@@ -114,16 +115,30 @@ def main():
     # old versions did not save class info in checkpoints, this walkaround is
     # for backward compatibility
-    if 'CLASSES' in checkpoint['meta']:
+    if 'CLASSES' in checkpoint.get('meta', {}):
```
Why might checkpoint not have meta?

LeonidBeynenson replied:
It happened once earlier during my experiments. It happened if checkpoint contains only a model without other data/metadata.
It seems to me that other mmcv code allows lo loading such checkpoints, so I kept the code after my experiments too.

LeonidBeynenson replied:
@Ilya-Krylov:  If you think that it may cause error during merging mmdetection updates later, I could remove that.

Ilya-Krylov replied:
It is hard to imagine that some pretrained mmdetection model cannot be run trough test.py script. If there is no reason to do this modification please don't do it. Return it back. As I said earlier we don't need extra changes in order not to complicate possible merging upstream issues.

LeonidBeynenson replied:
Ok. Fixed.

Ilya-Krylov commented on tools/test.py:
```diff
@@ -114,16 +115,30 @@ def main():
+    if cfg.get('nncf_config'):
+        check_nncf_is_enabled()
+        if not is_checkpoint_nncf(args.checkpoint):
+            raise RuntimeError('Trying to make testing with NNCF compression a model snapshot that was trained with NNCF')
```
Looks like it should be rephrased

LeonidBeynenson replied:
Sorry, `not` was missed. It was a typo.

LeonidBeynenson replied:
Fixed

kshpv commented on configs/nncf_compression/ssd/ssd300_coco_int8.py:
```diff
@@ -0,0 +1,26 @@
+load_from = 'http://download.openmmlab.com/mmdetection/v2.0/ssd/ssd300_coco/ssd300_coco_20200307-a92d2092.pth'
+
+find_unused_parameters = True
+
```
please add `total_epochs `value

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on mmdet/models/detectors/base.py:
```diff
@@ -1,16 +1,16 @@
+from contextlib import contextmanager
+from functools import partial
 from mmcv.utils import print_log
+from nncf.dynamic_graph.context import no_nncf_trace
```
import error in case of not installed nncf

LeonidBeynenson replied:
Fixed

LeonidBeynenson committed:
Make changes for nncf 1.4.1

LeonidBeynenson committed:
Fix several files w.r.t. PR comments

Mostly decorative changes and changes in documentation.
Also
* fix load_from in one NNCF config
* fix handling exceptions if unknown version of NNCF is imported

LeonidBeynenson committed:
Fix w.r.t. PR comments

Particularly
* fix import of no_nncf_trace
* fix documentation and comments

LeonidBeynenson commented:
2 @Ilya-Krylov:

>Where is it done? And looks like you load checkpoint twice. One time is here, and one more time is in wrap_nncf_model. Also checkpoint is loaded two times in test.py. Is it possible to wrap_nncf_model without loading weights and load weights after in current mmdetection manner?

Yes, we need to load checkpoints twice. First of all, we need to load the model with weights before wrapping in NNCF when we initialize NNCF inner variables (i.e. when NNCF fine-tuning is started). And even if we load a model that was pre-trined by NNCF I'm not sure if NNCF inner tricks work correctly if we do not load real weights before.

LeonidBeynenson committed:
Implement EvalBeforeRunHook and DistEvalBeforeRunHook

Earlier we added a new parameter to EvalHook/DistEvalHook for that, but
we decided to keep the classic mmdet hooks "as is" and implement our own
hook-s for evaluation before run.

LeonidBeynenson commented:
2 @Ilya-Krylov @druzhkov-paul :     
As far as I can see, I made all fixes w.r.t. most of your comments.    
(And wrote answers/explanations for all remaining comments)

Could you re-review this pull request again?

LeonidBeynenson committed:
Add `total_epochs` to nncf configs

Ilya-Krylov commented:
@LeonidBeynenson 

This branch is out-of-date with the base branch
Merge the latest changes from ote into this branch.

LeonidBeynenson committed:
Merge branch 'ote' into lb/merge_NNCF_support_to_ote

LeonidBeynenson committed:
Remove unrequired workaround on the rare case (meta not in mmdet's checkpoints)

LeonidBeynenson commented:
> @LeonidBeynenson
> 
> This branch is out-of-date with the base branch
> Merge the latest changes from ote into this branch.

Done

Ilya-Krylov commented:
run tests

Ilya-Krylov commented:
run tests

Ilya-Krylov commented:
@LeonidBeynenson unfortunately delivering CI statuses does not work at the moment. Please provide proof of green CI manually. Now it is red, but it is a kind of known issue (non-deterministic), probably not related to your PR http://134.191.240.124/job/ote/job/precommit/job/mmdetection/181/console

LeonidBeynenson commented:
> @LeonidBeynenson unfortunately delivering CI statuses does not work at the moment. Please provide proof of green CI manually. Now it is red, but it is a kind of known issue (non-deterministic), probably not related to your PR http://134.191.240.124/job/ote/job/precommit/job/mmdetection/181/console

2 @Ilya-Krylov :   
Unfortunately, I do have read-only access to this Jenkins. (My login/password do not work)  
So, I cannot restart CI.

LeonidBeynenson commented:
run tests

LeonidBeynenson commented:
Ilya, as far as I see, the pipeline is passed

http://134.191.240.124/job/ote/job/precommit/job/mmdetection/183/console

```
20:00:57  [Pipeline] End of Pipeline
20:00:57  Setting status of daa5c1fe7f28d0edfc13845c3fda21e750394844 to SUCCESS with url http://134.191.240.124/job/ote/job/precommit/job/mmdetection/183/ and message: 'Build finished. '
20:00:57  Using context: CI STATUS
20:00:57  Could not update commit status of the Pull Request on GitHub.
20:00:57  org.kohsuke.github.HttpException: {"message":"Bad credentials. The API can't be accessed using username/password authentication. Please create a personal access token to access this endpoint: 
...
20:00:57  	... 23 more
20:00:57  Finished: SUCCESS
```

Ilya, I think it is a SUCCESS, isn't it?

Commit:
Add a hack to fix saturation issue on AVX2 AVX512 (#3)

Add the function that tries to find the first convolution in backbone and
changes its export function to make export in int7 (instead of int8)
to avoid the saturation issue.

At the moment the trick works on Mobilenet-s only, for other backbones
it changes nothing

LeonidBeynenson commented:
run tests

LeonidBeynenson commented:
run tests

LeonidBeynenson commented:
@Ilya-Krylov :    
After update from Aleksei Kashapov re-run the tests again:  
http://134.191.240.124/job/ote/job/precommit/job/mmdetection/185/console  
This run returns "SUCCESS" again.

Ilya, please, note that update from Aleksei Kashapov changes only inner part of `mmdet/integration/nncf/compression.py`  
(it applies the changes to fix the saturation issue on AVX512 for Mobilenet-s)

Ilya-Krylov commented on mmdet/apis/train.py:
```diff
@@ -140,8 +160,16 @@ def train_detector(model,
         eval_hook = DistEvalHook if distributed else EvalHook
         runner.register_hook(eval_hook(val_dataloader, **eval_cfg))
 
+        if nncf_enable_compression:
```
I thought that you create EvalBeforeRunHook and DistEvalBeforeRunHook with the only before_run overridden method and make selection at line 160 depending on whether nncf enabled or not. That's it.

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on mmdet/integration/nncf/compression.py:
```diff
@@ -0,0 +1,210 @@
+        kwargs = {k: v.data[0] for k, v in dataloader_output.items()}
+        return (), kwargs
+
+    # TODO: not tested; need to test
```
have you tested?

kshpv replied:
yes. Actually this method is never used. It can be removed

LeonidBeynenson replied:
Fixed. Thanks Aleksei for change

Commit:
Remove non necessary function (#4)

LeonidBeynenson committed:
Change hooks to evaluate NNCF model before_run

Ilya-Krylov commented:
run tests
